src/charx.js

```diff
@@ -0,0 +1,415 @@
+        return mainIcon || iconAssets[0];
+    }
+
+    /**
+     * Normalize asset name for filesystem storage.
+     * @param {string} name - Original asset name
+     * @param {string} fallback - Fallback name if normalization fails
+     * @param {boolean} useHyphens - Use hyphens instead of underscores (for sprites)
+     * @returns {string} Normalized filename base (without extension)
+     */
+    getCharXAssetBaseName(name, fallback, useHyphens = false) {
+        const cleaned = (String(name ?? '').trim() || '');
+        if (!cleaned) {
+            return fallback.toLowerCase();
+        }
+
+        const separator = useHyphens ? '-' : '_';
+        // Convert to lowercase, collapse non-alphanumeric runs to separator, trim edges
+        const base = cleaned
+            .toLowerCase()
+            .replace(/[^a-z0-9]+/g, separator)
+            .replace(new RegExp(`^${separator}|${separator}$`, 'g'), '');
+
+        if (!base) {
+            return fallback.toLowerCase();
+        }
+
+        const sanitized = sanitize(base);
+        return (sanitized || fallback).toLowerCase();
+    }
+
+    mapCharXAssetsForStorage(assets) {
+        return assets.reduce((acc, asset) => {
+            if (!asset?.zipPath) {
+                return acc;
+            }
+
+            const ext = (asset.ext || '').toLowerCase();
+            if (!CHARX_IMAGE_EXTENSIONS.has(ext)) {
+                return acc;
+            }
+
+            if (asset.type === 'icon' || asset.type === 'user_icon') {
+                return acc;
+            }
+
+            let storageCategory;
+            if (CHARX_SPRITE_TYPES.has(asset.type)) {
+                storageCategory = 'sprite';
+            } else if (CHARX_BACKGROUND_TYPES.has(asset.type)) {
+                storageCategory = 'background';
+            } else {
+                storageCategory = 'misc';
+            }
+
+            // Use hyphens for sprites so ST's expression label extraction works correctly
+            // (sprites.js extracts label via regex that splits on dash or dot)
+            const useHyphens = storageCategory === 'sprite';
+            // Strip trailing extension from name if present (e.g., "image.png" with ext "png")
+            const nameWithoutExt = this.stripTrailingImageExtension(asset.name, ext);
+            acc.push({
+                ...asset,
+                ext,
+                storageCategory,
+                baseName: this.getCharXAssetBaseName(nameWithoutExt, `${storageCategory}-${asset.order ?? 0}`, useHyphens),
+            });
+
+            return acc;
+        }, []);
+    }
+}
+
+/**
+ * Delete existing file with same base name (any extension) before overwriting.
+ * Matches ST's sprite upload behavior in sprites.js.
+ * @param {string} dirPath - Directory path
+ * @param {string} baseName - Base filename without extension
+ */
+function deleteExistingByBaseName(dirPath, baseName) {
+    try {
+        const files = fs.readdirSync(dirPath);
```

Comment:
As I understand. This will also attempt to unlink directories with the same name. Here's a minimal fix.
```suggestion
        const files = fs.readdirSync(dirPath, { withFileTypes: true }).filter(f => f.isFile()).map(f => f.name);
```

Reply:
Thanks, made the suggested changes.